src/test/java/se/sundsvall/billingpreprocessor/service/InvoiceFileServiceTest.java

```diff
@@ -0,0 +1,309 @@
+
+		assertThat(billingRecordArgumentCaptor.getValue()).usingRecursiveComparison().ignoringFields("status").isEqualTo(entity);
+		assertThat(billingRecordArgumentCaptor.getValue().getStatus()).isEqualTo(INVOICED);
+		assertThat(invoiceFileArgumentCaptor.getValue()).satisfies(fileEntity -> {
+			assertThat(fileEntity.getType()).isEqualTo(INTERNAL.name());
+			assertThat(fileEntity.getName()).isEqualTo(FILENAME);
+			assertThat(fileEntity.getContent()).isEqualTo(new String(ArrayUtils.addAll(FILE_HEADER, INVOICE_DATA), StandardCharsets.UTF_8));
+			assertThat(fileEntity.getStatus()).isEqualTo(GENERATED);
+		});
+	}
+
+	@Test
+	void createBillingFilesWhenInvalidEntitiesExists() throws Exception {
+		// Arrange
+		final var internalFileName = "internalFileName";
+		final var externalFileName = "externalFileName";
+		final var internalFileHeader = "internalFileHeader".getBytes();
+		final var externalFileHeader = "externalFileHeader".getBytes();
+		final var internalInvoiceData = "internalInvoiceData".getBytes();
+		final var externalInvoiceData = "externalInvoiceData".getBytes();
+		final var invalidInternalEntity = createBillingRecordEntity(randomUUID().toString(), INTERNAL);
+		final var internalEntity = createBillingRecordEntity(randomUUID().toString(), INTERNAL);
+		final var invalidExternalEntity = createBillingRecordEntity(randomUUID().toString(), EXTERNAL);
+		final var externalEntity = createBillingRecordEntity(randomUUID().toString(), EXTERNAL);
+
+		when(billingRecordRepositoryMock.findAllByStatus(APPROVED)).thenReturn(List.of(
+			invalidInternalEntity, 
+			internalEntity,
+			invalidExternalEntity,
+			externalEntity));
+		when(invoiceFileConfigurationServiceMock.getInvoiceFileNameBy(INTERNAL.name(), CATEGORY)).thenReturn(internalFileName);
+		when(invoiceFileConfigurationServiceMock.getInvoiceFileNameBy(EXTERNAL.name(), CATEGORY)).thenReturn(externalFileName);
+		when(internalInvoiceCreatorMock.createFileHeader()).thenReturn(internalFileHeader);
+		when(internalInvoiceCreatorMock.createInvoiceData(internalEntity)).thenReturn(internalInvoiceData);
+		when(internalInvoiceCreatorMock.createInvoiceData(invalidInternalEntity)).thenThrow(Problem.valueOf(INTERNAL_SERVER_ERROR));
+		when(internalInvoiceCreatorMock.getProcessableTypes()).thenReturn(List.of(INTERNAL));
+		when(internalInvoiceCreatorMock.getProcessableCategories()).thenReturn(List.of(CATEGORY));
+		when(externalInvoiceCreatorMock.createFileHeader()).thenReturn(externalFileHeader);
+		when(externalInvoiceCreatorMock.createInvoiceData(externalEntity)).thenReturn(externalInvoiceData);
+		when(externalInvoiceCreatorMock.createInvoiceData(invalidExternalEntity)).thenThrow(Problem.valueOf(INTERNAL_SERVER_ERROR));
+		when(externalInvoiceCreatorMock.getProcessableTypes()).thenReturn(List.of(EXTERNAL));
+		when(externalInvoiceCreatorMock.getProcessableCategories()).thenReturn(List.of(CATEGORY));
```

Comment:
Isn't this a bit worrisome?

Reply:
I could have used real objects for this instead of a mock, but choose to use a mock in this case. The complexity comes from the test wanting to verify that even if entries with errors exists in the processed batch, a file with entries that does not have any errors will be created.

To make it less worrisome I will split the test into two (one for external file creation and one for internal file creation).